lib/src/sse_handler.dart

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:async/async.dart';
+import 'package:logging/logging.dart';
+import 'package:pedantic/pedantic.dart';
+import 'package:shelf/shelf.dart' as shelf;
+import 'package:stream_channel/stream_channel.dart';
+
+// RFC 2616 requires carriage return delimiters.
+String _sseHeaders(String origin) => 'HTTP/1.1 200 OK\r\n'
+    'Content-Type: text/event-stream\r\n'
+    'Cache-Control: no-cache\r\n'
+    'Connection: keep-alive\r\n'
+    'Access-Control-Allow-Credentials: true\r\n'
+    'Access-Control-Allow-Origin: $origin\r\n'
+    '\r\n\r\n';
+
+/// A bi-directional SSE connection between server and browser.
+class SseConnection extends StreamChannelMixin<String> {
+  /// Incoming messages from the Browser client.
+  final _incomingController = StreamController<String>();
+
+  /// Outgoing messages to the Browser client.
+  final _outgoingController = StreamController<String>();
+
+  Sink _sink;
+
+  /// How long to wait after a connection drops before considering it closed.
+  final Duration _keepAlive;
+
+  /// A timer counting down the KeepAlive period (null if connected).
+  Timer _keepAliveTimer;
+
+  /// The subscription that passes outgoing messages to the sink.
+  ///
+  /// This will be paused during the keepalive period and resumed upon reconnection.
+  StreamSubscription _outgoingStreamSubscription;
+
+  final _closedCompleter = Completer<void>();
+
+  /// Creates an [SseConnection] for the supplied [_sink].
+  ///
+  /// If [keepAlive] is supplied, the connection will remain active for this
+  /// period after a disconnect and can be reconnected transparently. If there
+  /// is no reconnect within that period, the connection will be closed normally.
+  ///
+  /// If [keepAlive] is not supplied, the connection will be closed immediately
+  /// after a disconnect.
+  SseConnection(this._sink, {Duration keepAlive}) : _keepAlive = keepAlive {
+    _outgoingStreamSubscription = _outgoingController.stream.listen((data) {
+      if (!_closedCompleter.isCompleted) {
+        try {
+          // JSON encode the message to escape new lines.
+          _sink.add('data: ${json.encode(data)}\n');
+          _sink.add('\n');
+        } catch (StateError) {
+          if (_keepAlive == null) {
+            rethrow;
+          }
+          // If we got here then the sink may have closed but the stream.onDone
+          // hasn't fired yet, so pause the subscription, re-queue the message
+          // and handle the error as a disconnect.
+          _handleDisconnect();
```

Comment:
Did you see my suggestion with using a StreamQueue here so that we can use `peek` and `next` to advance the queue only if we are successful? This way we don't have to re-queue the message.

Reply:
Yep, I replied to the comment here -> https://github.com/dart-lang/sse/pull/19#discussion_r359247315 I wasn't sure how best to change to it.

Reply:
I replied to the thread. Let me know if you have questions.